src/Mafia/Submodule.hs

```diff
@@ -103,9 +103,16 @@ getConfiguredSources = do
 
 getConventionSources :: EitherT GitError IO (Set Directory)
 getConventionSources = do
+  -- make sure we never include the current directory (i.e. the project we're
+  -- trying to build) as a source dependency
+  dir <- getCurrentDirectory
   root <- getProjectRoot
+  repoPaths <- Set.filter (/= dir) <$> getSourcesFrom root
+
   subs <- fmap ((root </>) . subName) <$> getSubmodules
-  Set.unions <$> traverse getSourcesFrom subs
+  subPaths <- Set.unions <$> traverse getSourcesFrom subs
+
+  return $ Set.union repoPaths subPaths
```

Comment:
If you try to make a circular dependency you get a relevant error message from cabal:
```
Checking for changes to dependencies...
Installing dependencies...
Resolving dependencies...
cabal: internal error: could not construct a valid install plan.
The proposed (invalid) plan contained the following problems:
The following packages are involved in a dependency cycle foo-0.0.1, bar-0.0.1, bar-test-0.0.1
-- snip --
```